### PR TITLE
Update 'Use pattern matching' to recognize other null checks

### LINKED ISSUE #36315: 'Use pattern matching' doesn't recognize 'is object' as a non-null check
**Version Used**: 16.2 Preview 1

**Steps to Reproduce**:

```csharp
int Method(object obj) {
  var str = obj as string;
  if (str is object) {
    return 1;
  }

  return 0;
}
```

**Expected Behavior**:

A code fix is offered to use pattern matching:

```csharp
int Method(object obj) {
  if (obj is string str) {
    return 1;
  }

  return 0;
}
```

**Actual Behavior**:

No code fix is offered.

### PR DESCRIPTION
**Pull request by sharwell**
* Recognize `is object` as a non-null check
* Recognize `is null` as a null check

Fixes #36315

### CODE REVIEW

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs [hunk 1]
```diff
@@ -89,18 +89,23 @@ void RemoveStatement(StatementSyntax statement)
...
 
             var declarator = (VariableDeclaratorSyntax)declaratorLocation.FindNode(cancellationToken);
-            var comparison = (BinaryExpressionSyntax)comparisonLocation.FindNode(cancellationToken);
+            var comparison = (ExpressionSyntax)comparisonLocation.FindNode(cancellationToken);
             var asExpression = (BinaryExpressionSyntax)asExpressionLocation.FindNode(cancellationToken);
+
+            var rightSideOfComparison =
+                comparison is BinaryExpressionSyntax binaryExpression
+                ? (SyntaxNode)binaryExpression.Right
+                : ((IsPatternExpressionSyntax)comparison).Pattern;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs]: for consistency with the rest of the code in the IDE, could you please write this either as:

```c#
            var rightSideOfComparison = comparison is BinaryExpressionSyntax binaryExpression
                ? (SyntaxNode)binaryExpression.Right
                : ((IsPatternExpressionSyntax)comparison).Pattern;

// or as

            var rightSideOfComparison =
                comparison is BinaryExpressionSyntax binaryExpression
                    ? (SyntaxNode)binaryExpression.Right
                    : ((IsPatternExpressionSyntax)comparison).Pattern;
```

Having deeper wrapping at the same nesting makes this pretty hard to read.

Thanks~

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -271,15 +280,34 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
 
             if (right.IsKind(SyntaxKind.NullLiteralExpression))
             {
+                // x == null
+                // x != null
+                return left;
+            }
+
+            if (right.IsKind(SyntaxKind.PredefinedType, out PredefinedTypeSyntax predefinedType)
+                && predefinedType.Keyword.IsKind(SyntaxKind.ObjectKeyword))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: what happens if someone writes `o == object`?  I suppose we just don't care about that sort of odd case?

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -271,15 +280,34 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
+            {
+                // x is object
+                return left;
+            }
+
+            if (right.IsKind(SyntaxKind.ConstantPattern, out ConstantPatternSyntax constantPattern)
+                && constantPattern.Expression is LiteralExpressionSyntax literalExpression
+                && literalExpression.IsKind(SyntaxKind.NullLiteralExpression))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: could simplify to `&& constantPattern.Expression.IsKind(NullLiteralExpression)`

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs (hunk 2 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: No idea. I'll think about ways to improve this.

Comment (sharwell) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Added checks for the comparison kind

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -271,15 +280,35 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
+                && predefinedType.Keyword.IsKind(SyntaxKind.ObjectKeyword)
+                && comparisonKind == SyntaxKind.IsExpression)
+            {
+                // x is object
+                return left;
+            }
+
+            if (right.IsKind(SyntaxKind.ConstantPattern, out ConstantPatternSyntax constantPattern)
+                && constantPattern.Expression.IsKind(SyntaxKind.NullLiteralExpression)
+                && comparisonKind == SyntaxKind.IsPatternExpression)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: 👍

**Pull request merged**